wakuv2/filter_manager.go

```diff
@@ -53,225 +28,95 @@ type FilterEvent struct {
-	}
-
 	return mgr
 }
-
-func (mgr *FilterManager) runFilterLoop(wg *sync.WaitGroup) {
-	defer wg.Done()
-	// Use it to ping filter peer(s) periodically
-	ticker := time.NewTicker(5 * time.Second)
-	defer ticker.Stop()
-
-	for {
-		select {
-		case <-mgr.ctx.Done():
-			mgr.logger.Debug("filter loop stopped")
-			return
-		case <-ticker.C:
-			mgr.pingPeers()
-		case ev := <-mgr.eventChan:
-			mgr.processEvents(&ev)
-		}
+func (mgr *FilterManager) isFilterSubAlive(ctx context.Context, sub *subscription.SubscriptionDetails) error {
+	if sub.Closed {
+		return errors.New("sub closed")
 	}
+	ctx, cancel := context.WithTimeout(ctx, pingTimeout)
+	defer cancel()
+	return mgr.node.IsSubscriptionAlive(ctx, sub)
 }
 
-func (mgr *FilterManager) processEvents(ev *FilterEvent) {
-	switch ev.eventType {
+func (mgr *FilterManager) filterLifecycle(ctx context.Context, filterID string, contentFilter protocol.ContentFilter) {
+	logger := mgr.logger.With(zap.String("filterID", filterID), zap.Any("contentFilter", contentFilter))
 
-	case FilterEventAdded:
-		mgr.filterSubs[ev.filterID] = make(subscription.SubscriptionSet)
-		mgr.resubscribe(ev.filterID)
-
-	case FilterEventRemoved:
-		for _, sub := range mgr.filterSubs[ev.filterID] {
-			if sub == nil {
-				// Skip temp subs
-				continue
-			}
-			go mgr.unsubscribeFromFilter(ev.filterID, sub)
-		}
-		delete(mgr.filterSubs, ev.filterID)
-
-	case FilterEventPingResult:
-		if ev.success {
-			break
-		}
-		// filterID field is only set when there are no subs to check for this filter,
-		// therefore no particular peers that could be unreachable.
-		if ev.filterID != "" {
-			// Trigger full resubscribe, filter has too few peers
-			mgr.logger.Debug("filter has too few subs", zap.String("filterId", ev.filterID))
-			mgr.resubscribe(ev.filterID)
-			break
-		}
+	// Use it to ping filter peer(s) periodically
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
 
-		// Delete subs for removed peer
-		for filterID, subs := range mgr.filterSubs {
-			for _, sub := range subs {
-				if sub == nil {
-					// Skip temp subs
-					continue
-				}
-				if sub.PeerID == ev.peerID {
-					mgr.logger.Debug("filter sub is inactive", zap.String("filterId", filterID), zap.Stringer("peerId", sub.PeerID), zap.String("subID", sub.ID))
-					delete(subs, sub.ID)
-					go mgr.unsubscribeFromFilter(filterID, sub)
-				}
-			}
-			mgr.resubscribe(filterID)
-		}
+	// A single sub for this filter
+	var sub *subscription.SubscriptionDetails
 
-	case FilterEventSubscribeResult:
-		subs, found := mgr.filterSubs[ev.filterID]
-		if ev.success {
-			if found {
-				subs[ev.sub.ID] = ev.sub
-				go mgr.runFilterSubscriptionLoop(ev.sub)
+	for {
+		logger.Debug("filter lifecycle started")
+		if sub != nil {
+			logger.Debug("filter health check")
+			err := mgr.isFilterSubAlive(ctx, sub)
+			if err != nil {
+				logger.Debug("filter ping error", zap.Error(err))
+				mgr.unsubscribe(ctx, logger, sub)
+				sub = nil
```

Comment:
Wondering it this identification of subscription not being alive should trigger some mechanism to fetch messages that could have been lost until we subscribe again for this Filter either via mailserver logic. 

Or will peer-syncing take care of such scenarios?